### PR TITLE
Move globals toName|toString to Tools: Tools.[getName|getString]

### PR DESCRIPTION
**Pull request by Slayer95**
- Requires discussion: <strike>Usage of ``Tools.getId`` vs ``this.getId`` / ``this.battle.getId`` in ``battle-engine.js``, ``data/*.js``, ``mods/**/*.js`` files.</strike>. Scrapped ``Tools.getId``.
- <strike>Also requires discussion: ``toName`` in ``team-validator.js``.</strike> Straightforwardly fixed.

### CODE REVIEW

**Comment by Zarel** [comment]
Why are you combining this with a different refactor?

Also, for now I think we should keep `getId` as a global because it's used everywhere.

**Comment by Slayer95** [comment]
Because it's a refactor needed for this refactor :p

Ok, will revert ``toId``

**Comment by Zarel** [comment]
I still don't understand why anything would be required for a simple function renaming.

**Comment by Slayer95** [comment]
Hmm you are right. It was intended to keep support of ``toId`` in child processes without loading the whole ``Tools`` module; however, in the end it wasn't actually needed.

I will just move the first part to a different PR after this is done with -which should be faster.

**Comment by Slayer95** [comment]
Updated

File: app.js [hunk 1]
```diff
@@ -301,45 +301,8 @@ global.toId = function (text) {
...
- * Safely ensures the passed variable is a string
- * Simply doing '' + str can crash if str.toString crashes or isn't a function
- * If we're expecting a string and being given anything that isn't a string
- * or a number, it's safe to assume it's an error, and return ''
- */
-global.string = function (str) {
-	if (typeof str === 'string' || typeof str === 'number') return '' + str;
-	return '';
+	if (typeof text !== 'string' && typeof str !== 'number') return '';
+	return ('' + text).toLowerCase().replace(/[^a-z0-9]+/g, '');
```
Comment (Slayer95) [inline app.js]: We can no longer call ``string``/``Tools.getString`` from ``toId`` or bad things happen

File: team-validator.js [hunk 2]
```diff
@@ -202,6 +178,19 @@ if (!process.send) {
 	});
 }
 
+/**
+ * Validates a username or Pokemon nickname
+ */
+
+var bannedNameStartChars = {'~':1, '&':1, '@':1, '%':1, '+':1, '-':1, '!':1, '?':1, '#':1, ' ':1};
+var toName = function (name) {
```
Comment (Zarel) [inline team-validator.js]: Why don't we use `getName` here?

Comment (Slayer95) [inline team-validator.js]: You mean the function name?

I just didn't find enough of a reason to rename it.

Comment (Slayer95) [inline team-validator.js]: Updated. Also switched to function declaration because I can (``toName`` used to be the result of an IIFE).

File: team-validator.js [hunk 3]
```diff
@@ -202,6 +178,19 @@ if (!process.send) {
...
 }
 
+/**
+ * Validates a username or Pokemon nickname
+ */
+
+var bannedNameStartChars = {'~':1, '&':1, '@':1, '%':1, '+':1, '-':1, '!':1, '?':1, '#':1, ' ':1};
+function getName (name) {
+	name = Tools.getName(name);
+	while (bannedNameStartChars[name.charAt(0)]) {
```
Comment (Zarel) [inline team-validator.js]: We might as well do `name.charAt(0) in bannedNameStartChars`.

Wait, why is this here? This should be in `Tools.getName`...

Comment (Zarel) [inline team-validator.js]: Or right below `namefilter`. This really only matters to usernames, not Pokémon names.

Comment (Slayer95) [inline team-validator.js]: Looks like https://github.com/Zarel/Pokemon-Showdown/commit/70ea1533a49198aa1c157e4d30897283182161b1 wasn't applied to team validator. Will do it. And we are left with just `Tools.getName``!

**Comment by Slayer95** [comment]
Updated

File: team-validator.js [hunk 4]
```diff
@@ -287,16 +263,16 @@ Validator = (function () {
 			return ["This is not a Pokemon."];
 		}
 
-		var template = tools.getTemplate(string(set.species));
+		var template = tools.getTemplate(Tools.getString(set.species));
```
Comment (Zarel) [inline team-validator.js]: At this point, `getString` is no longer necessary in the validator, I believe, since packed teams aren't raw JSON.

Well, first we need to drop support for non-packed teams, if we haven't already.

Comment (Zarel) [inline team-validator.js]: We should probably save that for a different pullreq, though.

File: tools.js [hunk 5]
```diff
@@ -221,6 +221,46 @@ module.exports = (function () {
...
+	 * If no such string can be found, returns the empty string. Calling
+	 * functions are expected to check for that condition and deal with it
+	 * accordingly.
+	 *
+	 * getName also enforces that there are not multiple space characters
+	 * in the name, although this is not strictly necessary for safety.
+	 */
+
+	Tools.prototype.getName = function (name) {
+		if (typeof name !== 'string' && typeof name !== 'number') return '';
```
Comment (Zarel) [inline tools.js]: I thought this worked like `toId`, where we'd extract `.name` from the object if it was an object?

Comment (Zarel) [inline tools.js]: Otherwise, perhaps this should be named `sanitizeName`?

Comment (Slayer95) [inline tools.js]: Changing it to extract ``.name`` would be cool.

File: team-validator.js (hunk 4 above)
Comment (Slayer95) [inline team-validator.js]: JSON teams haven't been supported for a while.

File: tools.js (hunk 5 above)
Comment (Zarel) [inline tools.js]: If it's currently unnecessary, though, we might as well keep it that way. I'd rather rename to `sanitizeName`.
